test/unit/test_tpm2_errata.c

```diff
@@ -0,0 +1,201 @@
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdarg.h>
+#include <stddef.h>
+
+#include <stdio.h>
+
+#include <setjmp.h>
+#include <cmocka.h>
+
+#include <sapi/tpm20.h>
+
+#include "tpm2_errata.h"
+#include "tpm2_util.h"
+
+static inline void setcaps(UINT32 level, UINT32 rev, UINT32 day, UINT32 year, TPM_RC rc) {
+
+    will_return(__wrap_Tss2_Sys_GetCapability, level);
+    will_return(__wrap_Tss2_Sys_GetCapability, rev);
+    will_return(__wrap_Tss2_Sys_GetCapability, day);
+    will_return(__wrap_Tss2_Sys_GetCapability, year);
+    will_return(__wrap_Tss2_Sys_GetCapability, rc);
+
+}
+
+TPM_RC __wrap_Tss2_Sys_GetCapability(TSS2_SYS_CONTEXT *sysContext,
+        TSS2_SYS_CMD_AUTHS const *cmdAuthsArray, TPM_CAP capability,
+        UINT32 property, UINT32 propertyCount, TPMI_YES_NO *moreData,
+        TPMS_CAPABILITY_DATA *capabilityData, TSS2_SYS_RSP_AUTHS *rspAuthsArray) {
+
+    UNUSED(sysContext);
+    UNUSED(cmdAuthsArray);
+    UNUSED(capability);
+    UNUSED(property);
+    UNUSED(propertyCount);
+    UNUSED(moreData);
+    UNUSED(capabilityData);
+    UNUSED(rspAuthsArray);
+
+    TPML_TAGGED_TPM_PROPERTY *properties = &capabilityData->data.tpmProperties;
+
+    properties->count = 4;
+
+    properties->tpmProperty[0].property = TPM_PT_LEVEL;
+    properties->tpmProperty[0].value    = (UINT32) mock();
+
+    properties->tpmProperty[1].property = TPM_PT_REVISION;
+    properties->tpmProperty[1].value    = (UINT32) mock();
+
+    properties->tpmProperty[2].property = TPM_PT_DAY_OF_YEAR;
+    properties->tpmProperty[2].value    = (UINT32) mock();
+
+    properties->tpmProperty[3].property = TPM_PT_YEAR;
+    properties->tpmProperty[3].value    = (UINT32) mock();
+
+    return (int) mock(); /* dequeue second value */
+}
+
+static void test_tpm2_errata_no_init_and_apply(void **state) {
+    UNUSED(state);
+
+    TPM2B_PUBLIC in_public = {
+        .publicArea = {
+            .objectAttributes = {
+                .val = TPMA_OBJECT_SIGN
+            },
+        },
+    };
+    tpm2_errata_fixup(SPEC_116_ERRATA_2_7,
+                      &in_public.publicArea.objectAttributes);
+
+    assert_int_equal(in_public.publicArea.objectAttributes.sign, 1);
+
+}
+
+static void test_tpm2_errata_bad_init_and_apply(void **state) {
+    UNUSED(state);
+
+    setcaps(00, 116, 303, 2014, TPM_RC_FAILURE);
+    tpm2_errata_init((TSS2_SYS_CONTEXT *) 0xDEADBEEF);
+
+
+    TPM2B_PUBLIC in_public = {
+        .publicArea = {
+            .objectAttributes = {
+                .val = TPMA_OBJECT_SIGN
+            },
+        },
+    };
+    tpm2_errata_fixup(SPEC_116_ERRATA_2_7,
+                      &in_public.publicArea.objectAttributes);
+
+    assert_int_equal(in_public.publicArea.objectAttributes.sign, 1);
+
```

Comment:
Inside errata code, the errata_match() will be true if this_errata_info is older than the known errata SPEC_116_ERRATA_2_7, where this_errata_info is retrieved from the tpm hardware. So I suspect setcaps() cannot affect this_errata_info.

So a tpm manufactured according to the original 1.16 spec will trigger the assertion because sign is fixed up to be 0.